tests/main/user-session-env/task.yaml

```diff
@@ -7,32 +7,53 @@ details: |
     inside the user session, no matter the shell they use.
 
 systems:
-   - -ubuntu-core-*  # cannot install zsh
+   - -ubuntu-core-*  # cannot install zsh or fish
    - -ubuntu-14.04-* # cannot use systemd
+   - -amazon-linux-2-* # no fish package for AMZN2
+
 
 environment:
     TEST_ZSH_USER: test-zsh
+    TEST_FISH_USER: test-fish
 
 prepare: |
+
     echo "Create a user with a different shell"
-    useradd --create-home --user-group -s /bin/zsh "$TEST_ZSH_USER"
-    tests.session prepare -u test,test-zsh
+    useradd --create-home --user-group -s /usr/bin/zsh "$TEST_ZSH_USER"
+    useradd --create-home --user-group -s /usr/bin/fish "$TEST_FISH_USER"
+    # tests.session assumes that the shell is sh compatible, which isn't true
+    # for fish
+    for user in test "$TEST_ZSH_USER" ; do
+        tests.session prepare -u "$user"
+    done
 
 restore: |
-    tests.session restore -u test,test-zsh
+    for user in test "$TEST_ZSH_USER"  ; do
+        tests.session restore -u "$user"
+    done
     userdel -f -r "$TEST_ZSH_USER"
+    userdel -f -r "$TEST_FISH_USER"
 
 execute: |
-    for user in test "$TEST_ZSH_USER"; do
+    for user in test "$TEST_ZSH_USER" ; do
         # Dump the environment set up by the user session manager
         if tests.session has-session-systemd-and-dbus; then
             tests.session -u "$user" exec systemctl --user show-environment > "${user}-session-env"
         fi
         tests.session -u "$user" exec env > "${user}-profile-env"
+        # fish allows environment variables to not be exported to subprocesses, make
+        # sure the right variables are exported though
+        tests.session -u "$user" exec sh -c 'exec env' > "${user}-child-env"
```

Comment:
This comment about fish is a bit misleading here: maybe rephrase it something like

> Later we will later check that the subprocess has all the variables it needs (not all shells behave in the same way)

And then you have anyway the specific comment about fish below.